### DISCUSSION TITLE
Different OAuth2 flows per route ?

### DISCUSSION THREAD

**Discussion by Abdelgha-4** [fastapi/fastapi#11616] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Description

The [OAuth2 scopes](https://fastapi.tiangolo.com/advanced/security/oauth2-scopes/) section in FastAPI Docs clearly explains how it is possible to use multiple `SecurityScopes` per routes as necessary. However, for my use-case I need to define different OAuth2 flows as well, basically something like this:

```python
oauth2_scheme = OAuth2(
    flows=OAuthFlows(
        implicit=OAuthFlowImplicit(
            authorizationUrl="",
            scopes={"scope0": "Common scope", "scope1": "Implicit only scope"},
        ),
        clientCredentials=OAuthFlowClientCredentials(
            tokenUrl="",
            scopes={"scope0": "Common scope", "scope2": "ClientCredentials only scope"},
        ),
    )
)

async def get_current_user(
    security_scopes: SecurityScopes, token: Annotated[str, Depends(oauth2_scheme)]
):
    ...

async def get_current_active_user(
    current_user: Annotated[User, Security(get_current_user, scopes=["scope1"], flows=["implicit"])],
):
    if current_user.disabled:
        raise HTTPException(status_code=400, detail="Inactive user")
    return current_user
```

`Security` Does not support a way to specify authorized `flows`. The scenario above for instance would enable only authorizing "implicit" users with the scope "scope1", and the general feature I'm looking for is specifying any combination of scopes and flows per route.

I thinks that one way of achieving this is building `oauth2_scheme` for each one of the desired combinations but that defeats the intended flexibly offered by `Security` and `SecurityScopes` so I hope there is a better solution.

### Operating System

Windows

### FastAPI Version

0.111.0

### Pydantic Version

2.7.1

### Python Version

3.11

**Comment by YuriiMotov** [comment]
Hi!
I think that there is no easy way (standard way) to do what you want.
I have one idea, but I need time to try it out. I will try to do it this weekend if nobody helps you

**Reply by YuriiMotov** [reply]
This is an explanation of my idea. I didn't manage to check it. So, it's just pseudo-code (not a working example):

```
from typing import Annotated

from fastapi import Depends, HTTPException, Security, status
from fastapi.openapi.models import OAuthFlowClientCredentials, OAuthFlowImplicit
from fastapi.security import OAuth2, OAuthFlows, SecurityScopes
from pydantic import BaseModel

IMPLICIT_FLOW_SCOPES = {
    "scope0": "Common scope",
    "implicit-scope1": "Implicit only scope",
}
CLIENT_CREDENTIALS_FLOW_SCOPES = {
    "scope0": "Common scope",
    "clientcredentials-scope2": "ClientCredentials only scope",
}

oauth2_scheme = OAuth2(
    flows=OAuthFlows(
        implicit=OAuthFlowImplicit(
            authorizationUrl="",
            scopes=IMPLICIT_FLOW_SCOPES,
        ),
        clientCredentials=OAuthFlowClientCredentials(
            tokenUrl="",
            scopes=CLIENT_CREDENTIALS_FLOW_SCOPES,
        ),
    )
)

class User(BaseModel):
    name: str

def decode_token(token: str):
    pass

async def get_user_from_db(user_id: str):
    pass

async def get_current_user(
    security_scopes: SecurityScopes, token: Annotated[str, Depends(oauth2_scheme)]
):
    implicit_scopes_required = [
        scope for scope in security_scopes if scope in IMPLICIT_FLOW_SCOPES
    ]
    client_credentials_scopes_required = [
        scope for scope in security_scopes if scope in CLIENT_CREDENTIALS_FLOW_SCOPES
    ]

    token = decode_token(token)

    if (not implicit_scopes_required) and (not client_credentials_scopes_required):
        return await get_user_from_db(token.sub)

    if implicit_scopes_required and set(implicit_scopes_required).issubset(
        token.scopes
    ):
        return await get_user_from_db(token.sub)

    if client_credentials_scopes_required and set(
        client_credentials_scopes_required
    ).issubset(token.scopes):
        return await get_user_from_db(token.sub)

    if implicit_scopes_required:
        implicit_scopes_required_str = (
            f"scopes = \"{' '.join(implicit_scopes_required)}\""
        )
    else:
        implicit_scopes_required_str = ""

    if client_credentials_scopes_required:
        client_credentials_scopes_required_str = (
            f"scopes = \"{' '.join(client_credentials_scopes_required)}\""
        )
    else:
        client_credentials_scopes_required_str = ""

    # Not sure about correctness of this,
    # but https://lists.w3.org/Archives/Public/ietf-http-wg/2012JanMar/0411.html
    www_authenticate_header = (
        f"Bearer {implicit_scopes_required_str}"
        f", , Bearer {client_credentials_scopes_required_str}"
    )

    raise HTTPException(
        status_code=status.HTTP_403_FORBIDDEN,
        detail="Not enough rights",
        headers={"WWW-Authenticate": www_authenticate_header},
    )

async def get_current_active_user(
    current_user: Annotated[
        User, Security(get_current_user, scopes=["implicit-scope1"])
    ],
):
    if current_user.disabled:
        raise HTTPException(status_code=400, detail="Inactive user")
    return current_user

```

The idea is:
1) include flow name into scope name (e.g. `implicit-scope1`)

2) before checking scopes, separate scopes of different flows (`implicit_scopes_required` and `client_credentials_scopes_required`)

3) user is authorized for operation if token's scopes match the required scopes of at least one of flows

Hope you will get the idea!

**Comment by Abdelgha-4** [comment]
Hi @YuriiMotov,
Thanks a lot for your answer!

Sorry, I think I was not very precise with my question. I actually did managed to get the access management logic working with "different flows and scopes per route", here is what my take on implementation looks like:

```python
def _parse_scopes(*scopes: str | Collection[str]) -> set[str]:
    return set(scopes.split(" ") if isinstance(scopes, str) else scopes)


def require_scopes_by_flow(
    implicit: str | Collection[str] | None = None,
    client: str | Collection[str] | None = None,
):
    assert implicit or client, "Must specify at least one flow"
    implicit = _parse_scopes(implicit)
    client = _parse_scopes(client)

    # the method get_token_info decodes the token and return its info as a pydantic model
    # including whether it's a client or implcit flow and the available scopes on the token
    # without relying on the 'cleaner' SecurityScopes
    def assert_scopes_present(
        token_info: Annotated[
            TokenInfoModel,
            Security(get_token_info, scopes=list(implicit | client)),
        ],
    ):
        if token_info.is_client_flow:
            if not client:
                raise HTTPException(
                    status_code=status.HTTP_403_FORBIDDEN,
                    detail="client flow is forbidden",
                )
            required_scopes = client
        else:
            if not implicit:
                raise HTTPException(
                    status_code=status.HTTP_403_FORBIDDEN,
                    detail="implicit flow is forbidden",
                )
            required_scopes = implicit

        missing_scopes = required_scopes - set(token_info.scope)
        if missing_scopes:
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail=f"Missing scopes for Authorization: {missing_scopes}",
            )

    return assert_scopes_present


async def get_current_active_user(
    current_user: Annotated[
        TokenInfoModel,
        Depends(require_scopes_by_flow(implicit="me", client=["read-only", "source"])),
    ],
):
    return current_user
```

The remaining issue is how to make the Swagger UI reflect correctly these rules, i.e when you click on Authorize on each endpoint you should only see valid choices for the Auth flow and the available scopes.

From a design point of view, sub-classing FastAPI's Security to support a new argument `flows` as a dictionary of flows -> scopes, and the same for maybe for `SecurityScopes` to detect the current used Auth flow looks like a neat solution. But I have no idea on the feasibility of this idea!

Another overkill non optimal solution would be to create different `oauth2_scheme` for each endpoint which would be then naively reflected on the SwaggerUI.

Thanks again for sharing your ideas on the issue!

**Comment by Abdelgha-4** [comment]
Closed in favor of the more precise #12090
